[CORE-950] Pipelines single account tutorial (#806)
gruntwork-io/docs: _docs-sources/pipelines/tutorial/index.md

```shell
terraform plan
```

Deploy the code using apply
```shell
terraform apply
```

## Step 3: Build and Push the Docker Images

The four standard Gruntwork Pipelines capabilities are instrumented by two separate docker files

1. [ecs-deploy-runner](./docker/deploy-runner/Dockerfile) - Terraform plan, apply and AMI building
2. [kaniko](./docker/kaniko/Dockerfile) - Docker image building. [Kaniko](https://github.com/GoogleContainerTools/kaniko) is a tool that supports building Docker images inside of a container

These Dockerfiles live in the ecs-deploy-runner module within [the terraform-aws-ci repository](https://github.com/gruntwork-io/terraform-aws-ci). In this example, we'll be cloning the terraform-aws-ci and running Docker build against the Dockerfiles defined there.

We're now going to build these two Docker images and push them to the ECR repositories we just created.

### Step 3a: Export Environment Variables

If you do not already have a GitHub Personal Access Token (PAT) available, you can follow this [guide to Create a new GitHub Personal Access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)

> For the purposes of this example, your token will need the `repo` scope, so that Gruntwork Pipelines is able to fetch modules and code from private Gruntwork repositories. Note that in production, the best practice would be to create a separate GitHub machine user account,

Reviewer: ```suggestion
For the purposes of this example, your token will need the `repo` scope, so that Gruntwork Pipelines is able to fetch modules and code from private Gruntwork repositories. Note that in production, the best practice is to create a separate GitHub machine user account,
```

   of the ephemeral ECS task container.

For now, export a valid GitHub PAT like so, so that we can use it to build Docker images that fetch private code via GitHub:
```shell
export GITHUB_OAUTH_TOKEN=<your-github-pat>
```

Export your AWS Account ID and primary region. The commands in the rest of this document require these variables to be set. The region to use is up to you.
```shell
export AWS_ACCOUNT_ID=<your-aws-account-id>
export AWS_REGION=<your-aws-region>
```

The Gruntwork Pipelines Dockerfiles used by Gruntwork Pipelines are stored in the `gruntwork-io/terraform-aws-ci` repository. Therefore, in order to pin both Dockerfiles
to a known version, we export the following variable which we'll use during our Docker builds:

```shell
export TERRAFORM_AWS_CI_VERSION=v0.51.4
```

The latest version can be retrieved from the [releases page](https://github.com/gruntwork-io/terraform-aws-ci/releases) of the `gruntwork-io/terraform-aws-ci` repository. At a minimum, `v0.51.4` must be selected.

### Step 3b: Clone `terraform-aws-ci` to your machine

> The following commands are going to use the Dockerfiles in this example, so first ensure that you have this repository cloned locally:

Reviewer: ```suggestion
Next, we are going to build the two Docker images required for this example. The Dockerfiles are defined in the [terraform-aws-ci](https://github.com/gruntwork-io/terraform-aws-ci) repository, so it must be available locally:
```

```shell
DOCKER_BUILDKIT=1 docker build \
  --secret id=github-token,env=GITHUB_OAUTH_TOKEN \
  --build-arg module_ci_tag="$TERRAFORM_AWS_CI_VERSION" \
  --tag "$AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/kaniko:$TERRAFORM_AWS_CI_VERSION" \
  ./docker/kaniko/
```

### Step 3d: Log and Push to ECR
Now we have local Docker images for ecs-deploy-runner and kaniko that are properly tagged, but before we can push it into the private ECR repository that we created
with our `terraform apply`, we need to authenticate with ECR itself. Authenticate to AWS and run the following:

```shell
aws ecr get-login-password --region $AWS_REGION \
  | docker login -u AWS --password-stdin "$AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com"
```

If you receive a success message from your previous command, you're ready to push your ecs-deploy-runner image:
```shell
docker push "$AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/ecs-deploy-runner:$TERRAFORM_AWS_CI_VERSION"
```

```shell
docker push "$AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/kaniko:$TERRAFORM_AWS_CI_VERSION"
```

## Step 4: Deploy the Pipelines Cluster

> Now that the ECR repositories are deployed and have the required docker images, we are ready

Reviewer: ```suggestion
Now that the ECR repositories are deployed and have the required Docker images, we are ready
```

> to deploy the rest of Gruntwork Pipelines. The terraform that defines the setup is defined in

Reviewer: ```suggestion
to deploy the rest of Gruntwork Pipelines. The Terraform that defines the setup is defined in
```

```shell
terraform apply
```

## Step 5: Install the `infrastructure-deployer` command line tool

Gruntwork Pipelines requires all requests to transit through its fronting Lambda function, which ensures only valid arguments and commands are passed along to ECS.
To invoke Gruntwork Pipelines's fronting Lambda function, you should use the `infrastructure-deployer` command line interface (CLI) tool. For testing and setup purposes, we'll install and use the `infrastructure-deployer` CLI locally; when you're ready to configure CI / CD, you'll install and use it in your CI / CD config.

If you do not already have the `gruntwork-install` binary installed, you can get it [here.](https://github.com/gruntwork-io/gruntwork-installer)

```bash

gruntwork-install --binary-name "infrastructure-deployer" --repo "https://github.com/gruntwork-io/terraform-aws-ci" --tag "$TERRAFORM_AWS_CI_VERSION"
```
:::note
If you'd rather not use the Gruntwork installer, you can alternatively download the binary manually from [the releases page.](https://github.com/gruntwork-io/terraform-aws-ci/releases)
:::

## Step 6: Invoke your Lambda Function

### Step 6a: Get your Lambda ARN from the output

> There is a Terraform output, `gruntwork_pipelines_lambda_arn`, configured that will tell you the ARN of the fronting Lambda that

Reviewer: ```suggestion
Next, we need to retrieve the Amazon Resource Name (ARN) for the Lambda function that guards your Gruntwork Pipelines installation.
```

I think it's a safe assumption that they missed it during apply time, so maybe we just say "retrieve it using the following command"?

 --ref "master" \
 --binary "terraform" \
 --command "plan"
```

If you forked the example repo provided you should see `+ out = "Hello, World"` if the plan was a success.

## Step 11: Celebrate, you did it!

As a next step you could add a `.github/workflows/pipeline.yml` file to your repo that runs the command above
or try it in your favorite CI/CD tool.  Your tooling only needs permission to trigger the lambda
function `arn:aws:lambda:us-east-1:<your account id>:function:ecs-deploy-runner-invoker`.

## Cleanup

If you want to remove the infrastructure created, you can use Terraform Destroy.

```shell
terraform plan -destroy -out terraform.plan
terraform apply terraform.plan
```

To destroy the `ecr-repositories` resources we created, we'll first need to empty the repos of any images:

```shell
aws ecr batch-delete-image --repository-name ecs-deploy-runner --image-ids imageTag=$TERRAFORM_AWS_CI_VERSION
aws ecr batch-delete-image --repository-name kaniko --image-ids imageTag=$TERRAFORM_AWS_CI_VERSION
aws ecr batch-delete-image --repository-name hello-world --image-ids imageTag=v1.0.0
```

> Then terraform can take care of the rest:

Reviewer: ```suggestion
Then Terraform can take care of the rest:
```
